### PR TITLE
Adding Dog-Owner appliction 

### LINKED ISSUE #26: Create DogOwner database and app
Create DogOwner database and app according to fields in ERD diagram .

### PR DESCRIPTION
**Pull request by tamirmatok**
Adding Dog-owner application. 
Features implemented: 
- creation of new dog-owner and user.

Dog-owner is exstending Django 'User' model .

Including: tests, migrations file, and data models .

closes #26

### CODE REVIEW

File: dogowner/models.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    dog_name = models.CharField(max_length=50)
+    dog_race = models.CharField(max_length=50)
+    dog_age = models.IntegerField()
+    dog_weight = models.FloatField()
+    dog_gender = models.CharField(
+        max_length=1,
+        choices=Gender,
+        default='M',)
+
+# should implement dog_picture
```
Comment (Omeramsc) [inline dogowner/models.py:24]: I suggest using an image url. don't try to store photos in the db, It'll only create a mess.

File: dogowner/models.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    Gender = (
+        ('M', 'Male'),
+        ('F', 'Female'),
+        )
+
+    email = models.EmailField(max_length=254)
+    first_name = models.CharField(max_length=50)
+    last_name = models.CharField(max_length=50)
+    user_name = models.CharField(max_length=50)
+    dog_name = models.CharField(max_length=50)
```
Comment (Omeramsc) [inline dogowner/models.py:15]: what happens when a Dogowner has more than one dog?

File: dogowner/models.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from django.db import models
+
+
+class Dogowner(models.Model):
```
Comment (Omeramsc) [inline dogowner/models.py:4]: Shouldn't it eventually inherit from User?

File: dogowner/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        migrations.CreateModel(
+            name='Dogowner',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True,
+                                           primary_key=True,
+                                           serialize=False,
+                                           verbose_name='ID')),
+                ('first_name', models.CharField(max_length=255)),
+                ('last_name', models.CharField(max_length=255)),
+                ('phone_number', models.CharField(max_length=20)),
```
Comment (OfirMatasas) [inline dogowner/migrations/0001_initial.py:26]: I think that max_length of phone_number should be 10, because there's no phone number that long, and to prevent typos.

File: dogowner/migrations/0001_initial.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                ('dog_picture_url', models.CharField(max_length=255)),
+                ('dog_age', models.IntegerField()),
+                ('dog_weight', models.FloatField()),
+                ('dog_gender', models.CharField(choices=[('M', 'Male'),
+                                                         ('F', 'Female'),
+                                                         ('UN', 'Unknown')],
+                                                default='M',
+                                                max_length=2)),
+                ('user', models.ForeignKey(blank=True,
+                                           null=True,
```
Comment (OfirMatasas) [inline dogowner/migrations/0001_initial.py:38]: Why does the foreign key of "user" accept null?
Is there a scenario where the dog owner doesn't have a user to rebase on?
Same goes to blank.

File: dogowner/models.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    Female = 'F', 'Female'
+    Unknown = 'UN', 'Unknown'
+
+
+class Dogowner(models.Model):
+
+    user = models.ForeignKey(User,
+                             on_delete=models.CASCADE,
+                             blank=True,
+                             null=True
```
Comment (OfirMatasas) [inline dogowner/models.py:16]: Same as my last commend in Migration class

File: pytest.ini [hunk 7]
```diff
@@ -1,4 +1,4 @@
 [pytest]
 DJANGO_SETTINGS_MODULE = Hotails.settings
 django_debug_mode = true
-python_files = tests.py test_*.py *_tests.py
+python_files = tests.py test_*.py *_tests.py *_test.py
```
Comment (OfirMatasas) [inline pytest.ini:4]: Jenia already took care of this bug in #44

**Review by OfirMatasas — CHANGES_REQUESTED** [review]
Looks good overall! Just a few things to take care of:
1. phone number max_length should be 10, to prevent typos.
2. User Foreign key should not accept blank and null values.
3. pytest.ini - Jenia took care of it in #44 

I also wrote those notes in the files you committed, to make it easier to find the code I'm referring to.

**Comment by tamirmatok** [comment]
> Looks good overall! Just a few things to take care of:
> 
> 1. phone number max_length should be 10, to prevent typos.
> 2. User Foreign key should not accept blank and null values.
> 3. pytest.ini - Jenia took care of it in #44
> 
> I also wrote those notes in the files you committed, to make it easier to find the code I'm referring to.

Tnx for reviewing @OfirMatasas 
1) NP I will change it .
2) I agree with you ,but it seems to be necessary because when I tried to makemigrations without it I got an error that I have to supply default value. 
3) I know , I branch out from his branch .

File: dogowner/models.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                             )
+    first_name = models.CharField(max_length=255)
+    last_name = models.CharField(max_length=255)
```
Comment (Omeramsc) [inline dogowner/models.py:19]: same

File: dogowner/models.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    phone_number = models.CharField(max_length=20)
```
Comment (Omeramsc) [inline dogowner/models.py:20]: we can assume it's max length is 10 digits.
it should probably also be integers.
we can ignore international numbers for now IMO (but even then, 20 Is too long)

File: dogowner/models.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+class Gender(models.TextChoices):
+    Male = 'M', 'Male'
+    Female = 'F', 'Female'
+    Unknown = 'UN', 'Unknown'
+
+
+class Dogowner(models.Model):
+
+    user = models.ForeignKey(User,
```
Comment (Omeramsc) [inline dogowner/models.py:13]: you should probably use `models.OneToOneField` instead.

File: dogowner/models.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                             on_delete=models.CASCADE,
+                             blank=True,
+                             null=True
+                             )
+    first_name = models.CharField(max_length=255)
```
[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/models.py:18]: I hope this person doesn't have a dog :wink:

File: dogowner/models.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        choices=Gender.choices,
+        default='M',)
+
+    DogOwners = models.Manager()
+
+    def __str__(self):
+        return self.first_name + self.last_name
+
+    @staticmethod
+    def create_dogowner(email, username, password, dog_name,
```
Comment (Omeramsc) [inline dogowner/models.py:37]: Why do you need this method?

File: dogowner/models.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                                 dog_picture_url=dog_picture_url,
+                                 dog_age=dog_age,
+                                 dog_weight=dog_weight,
+                                 dog_gender=dog_gender
+                                 )
+        new_dog_owner.save()
+        return new_dog_owner
+
+    @staticmethod
+    def get_user(username):
```
Comment (Omeramsc) [inline dogowner/models.py:61]: is this needed?

File: dogowner/models.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        try:
+            user = User.objects.get(username=username)
+        except User.DoesNotExist:
+            return None
+        return user
+
+    @staticmethod
+    def del_dog_owner(username):
```
Comment (Omeramsc) [inline dogowner/models.py:69]: is this needed now?

File: dogowner/test_dog_owner.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+    def test_persist_member(self, dog_owner_0):
+        dog_owner_0.save()
+        assert dog_owner_0 in Dogowner.DogOwners.all()
+
+    def test_get_user_from_dog_owner(self, dog_owner_0):
+        assert Dogowner.get_user(USERNAME) == dog_owner_0.user
+        assert Dogowner.get_user('no_user') is None
+
+    def test_del_dog_owner(self, dog_owner_0):
```
Comment (Omeramsc) [inline dogowner/test_dog_owner.py:56]: AFAIK Django already provides you with a deletion method. you don't need to use your own.

**Review by Omeramsc — CHANGES_REQUESTED** [review]

File: dogowner/models.py (hunk 12 above)
Comment (tamirmatok) [reply dogowner/models.py:37]: I had to find a way for init Django user with our requierd fields , and make sure we will not have a situation of creating dog- owner without user initalizion.
That is why I assumed this method can help .

File: dogowner/models.py (hunk 13 above)
Comment (tamirmatok) [reply dogowner/models.py:61]: I tried to provide this model with functions that support several actions,
In second thought, it is not nesscary in this point .
If I will see this method usful in the future, I will add it .

File: dogowner/models.py (hunk 14 above)
Comment (tamirmatok) [reply dogowner/models.py:69]: same

File: dogowner/models.py (hunk 12 above)
Comment (tamirmatok) [reply dogowner/models.py:37]: I will try to search for "cleaner" way for this task .

File: dogowner/test_dog_owner.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+DOG_RACE = 'lavrador'
+DOG_PICTURE_URL = 'http....'
+DOG_AGE = 10
+DOG_WEIGHT = 6
+DOG_GENDER = 'M'
+
+
+@pytest.fixture
+def dog_owner_0():
+    return Dogowner.create_dogowner(email=EMAIL,
```
[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/test_dog_owner.py:20]: After I will know your opinion on https://github.com/redhat-beyond/Hotails/pull/33#discussion_r833330154 I guess I will find a better way .

[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/test_dog_owner.py:20]: After i will now your opinion on https://github.com/redhat-beyond/Hotails/pull/33#discussion_r833330154 i guess i will know the right way .

File: dogowner/models.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/models.py]: Fixed

File: dogowner/models.py (hunk 8 above)
Comment (tamirmatok) [reply dogowner/models.py]: Fixed

File: dogowner/models.py (hunk 9 above)
Comment (tamirmatok) [reply dogowner/models.py]: Fixed

File: dogowner/models.py (hunk 10 above)
Comment (tamirmatok) [reply dogowner/models.py]: Fixed

File: dogowner/test_dog_owner.py (hunk 15 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py:56]: Fixed

File: dogowner/test_dog_owner.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/test_dog_owner.py]: Fixed .

File: dogowner/models.py (hunk 12 above)
Comment (tamirmatok) [reply dogowner/models.py]: According to https://github.com/redhat-beyond/Hotails/pull/33#discussion_r833328410 , I gave up on this method and we will use regular initializion.
I guess because we don't have any registration in our application ,we can make sure that "User" entity will initalize with our required fields .

**Review by ErezCohenn — APPROVED** [review]
Great Job @tamirmatok !
don't forget to add `Data migrations` (consider maybe in another PR)

**Review by OfirMatasas — APPROVED** [review]
Great work!

File: dogowner/models.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        max_length=2,
+        choices=Gender.choices,
+        default='UN', blank=True)
+
+    def __str__(self):
+        return self.first_name + self.last_name
+
+
+@receiver(post_save, sender=User)
+def create_user_profile(sender, instance, created, **kwargs):
```
Comment (Yuval-Vino) [inline dogowner/models.py:33]: so the"@reciver" was the magic solution for this? surely this will help us all

**Review by Yuval-Vino — APPROVED** [review]
Looks really good! Good job!!

File: dogowner/models.py (hunk 17 above)
Comment (tamirmatok) [reply dogowner/models.py:33]: Exactly, we use signal of post save for creation&saving Dog-owner after signal occurred.

File: dogowner/models.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+from django.dispatch import receiver
+
+
+class Gender(models.TextChoices):
+    Male = 'M', 'Male'
+    Female = 'F', 'Female'
+    Unknown = 'UN', 'Unknown'
+
+
+class Dogowner(models.Model):
```
Comment (jeniaSakirko) [inline dogowner/models.py:13]: The class name should be camelcase, see https://realpython.com/python-pep8/#naming-styles

File: dogowner/apps.py [hunk 19]
```diff
@@ -0,0 +1,6 @@
+from django.apps import AppConfig
+
+
+class DogownerConfig(AppConfig):
```
Comment (jeniaSakirko) [inline dogowner/apps.py:4]: The class name should be camelcase, see https://realpython.com/python-pep8/#naming-styles

**Review by jeniaSakirko — CHANGES_REQUESTED** [review]

File: dogowner/test_dog_owner.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+                                  dog_age=DOG_AGE,
+                                  dog_weight=DOG_WEIGHT,
+                                  dog_gender=DOG_GENDER
+                                  )
+                )
+
+
+@pytest.mark.django_db
+class TestDogOwnerModel:
+    def test_create_dogowner(self, dog_owner_0):
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:39]: The tests lack coverage of edge cases for example:
what happens if you try to add the same user twice?
Think of them and add more tests accordingly.

**Review by ormergi — COMMENTED** [review]
Hi Tamir,
Please squash intermediate commits.
I suggest to rethink how to represent the dog-owner in the code in a way 
it will be easy to test.

File: dogowner/models.py (hunk 17 above)
Comment (tamirmatok) [reply dogowner/models.py:33]: As I see it we have 3 options :
1) Create a function for saving 'User' with our requierd field and than saving 'Dog-Owner' fields - Omer told me it is not the right way for doing it.
2) Init 'User' in dog-owner creation, but than in tests i have to save/delete 'User' and 'Dog-Owner' seperatly, what make the creation/deletion more complex.
3) Listen to a signal and manage all the creation from 'User'  - what seems to me the most ellegant way .

I have took my resource from https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html
Please let me know what you think @ormergi

Comment (ormergi) [reply dogowner/models.py:33]: I find the "AbstractUser way" much more simple.

Comment (tamirmatok) [reply dogowner/models.py:33]: Ok thanks for reviewing . @ormergi

**Review by jeniaSakirko — APPROVED** [review]
LGTM

File: dogowner/test_dog_owner.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+                           dog_age=DOG_AGE,
+                           dog_weight=DOG_WEIGHT,
+                           dog_gender=DOG_GENDER
+                           )
+
+
+@pytest.mark.django_db
+class TestDogOwnerModel:
+    def test_persist_dog_owner(self, dogOwner_01):
+        dogOwner_01.save()
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:40]: save() is called on creation, why is it necessary to call again?

File: dogowner/test_dog_owner.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+PHONE_NUMBER = 1234567890
+DOG_RACE = 'lavrador'
+DOG_PICTURE_URL = 'https://www.google.com/'
+DOG_AGE = 10
+DOG_WEIGHT = 6
+DOG_GENDER = 'M'
+
+
+@pytest.fixture
+def dogOwner_01():
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:21]: Please rename this fixture per what it does, it should be understood by its name.

File: dogowner/test_dog_owner.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+        assert dogOwner_01.user in User.objects.all()
+        assert dogOwner_01 in DogOwner.objects.all()
+
+    def test_del_dogowner(self, dogOwner_01):
+        dogOwner_01.save()
+        dogOwner_01.delete()
+        assert dogOwner_01 not in User.objects.all()
+        assert dogOwner_01 not in DogOwner.objects.all()
+
+    def test_invalid_email(self, dogOwner_01):
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:50]: Consider rename this and other tests as well to a name that would describe the whole story,
something like "test_dog_owner_user_creation_with_invalid_email"

File: dogowner/test_dog_owner.py [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+        with pytest.raises(ValidationError, match="'Enter a valid email address.'"):
+            DogOwner.create(email='INVALID_EMAIL', username='testuser02', password=PASSWORD,
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:52]: Why this test calls `dogOwner_01` fixture?
Same for other tests in this file.

File: dogowner/test_dog_owner.py [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+                           dog_race=DOG_RACE,
+                           dog_picture_url=DOG_PICTURE_URL,
+                           dog_age=DOG_AGE,
+                           dog_weight=DOG_WEIGHT,
+                           dog_gender=DOG_GENDER
+                           )
+
+
+@pytest.mark.django_db
+class TestDogOwnerModel:
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:38]: Please add test to cover the case where DogOwner is created twice.

File: dogowner/models.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+from .validators import validate_url, validate_username, validate_phone
+
+
+class Gender(models.TextChoices):
+    Male = 'M', 'Male'
+    Female = 'F', 'Female'
+    Unknown = 'UN', 'Unknown'
+
+
+class DogOwner(models.Model):
```
Comment (ormergi) [inline dogowner/models.py:13]: Please consider changing DogOwner to inherit from the django builtin AbstrctUser / User,
it was discussed before and concluded to be the most simple solution.
It should simplify the code by using the builtin save() and create() functions,
as DogOwner will act as a facade of the builtin User, there should be no need for managing 
both DogOwner and User lifecycle (e.g: calling delete() or save()).

File: dogowner/models.py [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+    def __str__(self):
+        return self.first_name + self.last_name
+
+    @staticmethod
+    def create(email, username, password, dog_name,
+               first_name, last_name, phone_number,
+               dog_race, dog_picture_url, dog_age,
+               dog_weight, dog_gender):
+
+        # Validation
```
Comment (ormergi) [inline dogowner/models.py:37]: Please remove such comments, it can be understood by the code.

On a general note,  comments rot, if there is something that should be explain forwarder 
then what can be understood for the code, it should be documented at the commit message
or proper doc.

File: dogowner/test_dog_owner.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+                            dog_age=DOG_AGE, dog_weight=DOG_WEIGHT, dog_gender=DOG_GENDER
+                            )
+
+    def test_phone_long(self, dogOwner_01):
+        with pytest.raises(ValidationError, match="Invalid phone - phone should be 10 digits"):
+            DogOwner.create(email=EMAIL, username='invalid_phone', password=PASSWORD,
+                            dog_name=DOG_NAME, first_name=FIRST_NAME, last_name=LAST_NAME,
+                            phone_number=1234567890123, dog_race=DOG_RACE, dog_picture_url=DOG_PICTURE_URL,
+                            dog_age=DOG_AGE, dog_weight=DOG_WEIGHT, dog_gender=DOG_GENDER
+                            )
```
Comment (ormergi) [inline dogowner/test_dog_owner.py:97]: Why is it necessary to call dogOwner_01 fixture?

File: dogowner/models.py (hunk 26 above)
Comment (tamirmatok) [reply dogowner/models.py:13]: Hey @ormergi , 
While Inherit from AbstractUser , I don’t understand how can I solve https://github.com/redhat-beyond/Hotails/pull/33#discussion_r833330154 , we will still have to implement create method for this.
I think according to ERD diagram we should have User field inside DogOwner class , that is why i think OneToOne field is better solution .
In addition, the deletion is not a problem in this case because we mark 'on_delete=models.CASCADE' . 
If we dont want any seperation with DogOwner and User, it will be better to create DogOwner class from zero without exstanding anything , again for not dealing with undesirable fields .

File: dogowner/test_dog_owner.py (hunk 28 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py]: You right , fixed .

File: dogowner/models.py (hunk 27 above)
Comment (tamirmatok) [reply dogowner/models.py]: Fixed.

File: dogowner/models.py (hunk 26 above)
Comment (tamirmatok) [reply dogowner/models.py:13]: I have share my opinion on slack . Waiting for your review .

File: dogowner/test_dog_owner.py (hunk 24 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py:53]: Fixed

File: dogowner/test_dog_owner.py (hunk 23 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py]: Fixed.

File: dogowner/test_dog_owner.py (hunk 22 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py]: Fixed.

File: dogowner/test_dog_owner.py (hunk 21 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py]: You right. Fixed .

File: dogowner/models.py (hunk 26 above)
Comment (ormergi) [reply dogowner/models.py:13]: Please share your opinion here as well :slightly_smiling_face:

File: dogowner/test_dog_owner.py (hunk 25 above)
Comment (tamirmatok) [reply dogowner/test_dog_owner.py:39]: I have add a test checking same username creation , it is promise not to have same DogOwner on table .

File: dogowner/models.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+    @staticmethod
+    def create(email, username, password, dog_name,
+               first_name, last_name, phone_number,
+               dog_race, dog_picture_url, dog_age,
+               dog_weight, dog_gender):
+
+        validate_email(email)
+        validate_username(username)
+        validate_url(dog_picture_url)
+        validate_phone(phone_number)
```
Comment (ormergi) [inline dogowner/models.py:40]: Are other fields require validation? if so add it and cover it with test.

File: dogowner/migrations/0001_initial.py [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+                ('dog_race', models.CharField(blank=True, max_length=15)),
+                ('dog_picture_url', models.CharField(blank=True, max_length=1000)),
+                ('dog_age', models.IntegerField(blank=True, null=True)),
+                ('dog_weight', models.FloatField(blank=True, null=True)),
+                ('dog_gender', models.CharField(blank=True, choices=[('M', 'Male'),
+                                                                     ('F', 'Female'),
+                                                                     ('UN', 'Unknown')],
+                                                default='UN', max_length=2)),
+                ('user', models.OneToOneField(on_delete=django.db.models.deletion.CASCADE,
+                                              to=settings.AUTH_USER_MODEL)),
```
Comment (ormergi) [inline dogowner/migrations/0001_initial.py:37]: Why is ` to=settings.AUTH_USER_MODE` necessary?

Comment (ormergi) [reply dogowner/migrations/0001_initial.py:37]: I just saw that its generated by django

File: dogowner/test_dog_owner.py [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+PHONE_NUMBER = 1234567890
+DOG_RACE = 'lavrador'
+DOG_PICTURE_URL = 'https://www.google.com/'
+DOG_AGE = 10
+DOG_WEIGHT = 6
+DOG_GENDER = 'M'
+
+
+@pytest.fixture
+def create_dogowner_01():
```
[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/test_dog_owner.py]: Fixed.

File: dogowner/validators.py [hunk 32]
```diff
@@ -0,0 +1,22 @@
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+from django.core.validators import URLValidator
+
+
+def validate_username(username):
+    if User.objects.filter(username=username).exists():
+        raise ValidationError("Invalid username")
```
[the remark this answers is not in the document]
Comment (tamirmatok) [reply dogowner/validators.py]: Done.

File: dogowner/models.py (hunk 29 above)
Comment (tamirmatok) [reply dogowner/models.py]: Done.

**Review by ErezCohenn — APPROVED** [review]
You did a great job @tamirmatok! well done.

**Review by jeniaSakirko — COMMENTED** [review]
LGTM

**Review by OfirMatasas — APPROVED** [review]
LGTM

**Review by ormergi — APPROVED** [review]
Thank you, it looks good.
